### PR TITLE
[10.0][MIG] : account_invoice_customer_ref_unique

### PR DESCRIPTION
**Pull request by misern2**
# USAGE

This module adds a insensitive constraint on the name Customer Reference
(name field): (partner_id, name) must be unique.

### CODE REVIEW

**Review by AaronHForgeFlow — APPROVED** [review]
:+1: 
It'll be great a pre-install hook to prevent this constraint to be triggered when installing.

File: account_invoice_customer_ref_unique/models/account_invoice.py [hunk 1]
```diff
@@ -19,47 +19,34 @@
...
-            'name': '',
-        })
-        return super(AccountInvoice, self).copy(cr, uid, ids, default, context)
-
-    def _check_unique_name_insensitive(self, cr, uid, ids, context=None):
-        for i_id in ids:
-            invoice = self.browse(cr, uid, i_id, context=context)
+    @api.multi
+    @api.constrains('name')
+    def _check_unique_name_insensitive(self):
```
Comment (moylop260) [inline account_invoice_customer_ref_unique/models/account_invoice.py]: In order to increase the performance is faster use a sql constraint based on computed fields with store=True

Comment (moylop260) [reply account_invoice_customer_ref_unique/models/account_invoice.py]: check the following example where is 4x faster than a python constraint using too many invoices https://github.com/Vauxoo/addons-vauxoo/blob/8.0/supplier_invoice_number_unique/model/account_invoice.py

**Review by moylop260 — CHANGES_REQUESTED** [review]

**Comment by misern2** [comment]
@moylop260 Thanks for your revision. I pushed a new commit with the SQL Constraints

File: account_invoice_customer_ref_unique/models/account_invoice.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    def _compute_customer_ref_strip(self):
+        for invoice in self:
+            invoice_type = invoice.type
+            if invoice_type not in ['out_invoice', 'out_refund']:
+                return True
+            if invoice.name:
+                invoice.customer_ref_strip = invoice.name.lower()
+    _sql_constraints = [
+        ('unique_customer_ref_strip', 'UNIQUE('
+         'customer_ref_strip, company_id, partner_id)',
```
Comment (moylop260) [inline account_invoice_customer_ref_unique/models/account_invoice.py]: s/partner_id/commercial_partner_id/g` in order to supports the case where other address is used for the same partner

File: .travis.yml [hunk 3]
```diff
@@ -1,7 +1,7 @@
 language: python
 
 python:
-  - "2.7"
+  - "2.7"
```
Comment (moylop260) [inline .travis.yml]: remove this small change

**Review by moylop260 — CHANGES_REQUESTED** [review]
LGTM just small changes.

And check our beta message of the lint build:
https://travis-ci.org/OCA/account-invoicing/jobs/281607391#L743

```bash
************* Module account_invoice_customer_ref_unique
account_invoice_customer_ref_unique/README.rst:32: [E7901(rst-syntax-error), ]  Unexpected possible title overline or transition.|Treating it as ordinary text because it's so short.
account_invoice_customer_ref_unique/README.rst:1: [W7908(missing-newline-extrafiles), ]  Missing newline
************* Module account_invoice_customer_ref_unique.__manifest__
account_invoice_customer_ref_unique/__manifest__.py:1: [C8201(no-utf8-coding-comment), ] No UTF-8 coding comment found: Use `# coding: utf-8` or `# -*- coding: utf-8 -*-`
account_invoice_customer_ref_unique/__manifest__.py:23: [C8103(manifest-deprecated-key), ] Deprecated key "description" in manifest file
************* Module account_invoice_customer_ref_unique.models.account_invoice
account_invoice_customer_ref_unique/models/account_invoice.py:1: [C8201(no-utf8-coding-comment), ] No UTF-8 coding comment found: Use `# coding: utf-8` or `# -*- coding: utf-8 -*-`
```

**Comment by moylop260** [comment]
...and create a hook to auto-fix all current wrong info breaking the constraint:
https://travis-ci.org/OCA/account-invoicing/jobs/281607395#L1407

**Comment by misern2** [comment]
@moylop260 I think it's all finshed. I also checked the travis log file and now  there is not any error beta message.
Thanks.

File: account_invoice_customer_ref_unique/__init__.py [hunk 4]
```diff
@@ -20,4 +20,5 @@
 #
 ##############################################################################
 
-from . import account_invoice
+
+from . import models
```
Comment (moylop260) [inline account_invoice_customer_ref_unique/__init__.py]: Missing newline

File: account_invoice_customer_ref_unique/__init__.py [hunk 5]
```diff
@@ -20,4 +20,5 @@
 #
 ##############################################################################
```
Comment (moylop260) [inline account_invoice_customer_ref_unique/__init__.py]: Use shortest license like as template module:
https://github.com/OCA/maintainer-tools/blob/master/template/module/models/model_name.py#L2

File: account_invoice_customer_ref_unique/README.rst [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+   :alt: Odoo Community Association
+   :target: https://odoo-community.org
+
+This module is maintained by the OCA.
+
+OCA, or the Odoo Community Association, is a nonprofit organization whose
+mission is to support the collaborative development of Odoo features and
+promote its widespread use.
+
+To contribute to this module, please visit https://odoo-community.org.
```
Comment (moylop260) [inline account_invoice_customer_ref_unique/README.rst:59]: missing newline

File: account_invoice_customer_ref_unique/README.rst [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+Installation
+============
+
+To install this module, just selected from available modules.
+
+
+Configuration
+=============
+
```
Comment (moylop260) [inline account_invoice_customer_ref_unique/README.rst]: Remove unused sections

File: account_invoice_customer_ref_unique/models/account_invoice.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
+# -*- coding: utf-8 -*-
+# #############################################################################
```
Comment (moylop260) [inline account_invoice_customer_ref_unique/models/account_invoice.py]: Short license

**Review by moylop260 — CHANGES_REQUESTED** [review]
Sorry I didn't see this before...

Thanks for your great work!

File: account_invoice_customer_ref_unique/models/account_invoice.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+    customer_ref_strip = fields.Char(
+        compute='_compute_customer_ref_strip', store=True)
+
+    @api.multi
+    @api.depends('name')
+    def _compute_customer_ref_strip(self):
+        for invoice in self:
+            invoice_type = invoice.type
+            if invoice_type not in ['out_invoice', 'out_refund']:
```
Comment (houssine78) [inline account_invoice_customer_ref_unique/models/account_invoice.py]: maybe we can remove the previous line as the invoice_type variable is only used once in the if statement

**Comment by misern2** [comment]
Changes requested.
The travis build on my branch was Ok and It fails on this PR in my last commit. Am I missing something ?

**Review by houssine78 — APPROVED** [review]

**Comment by pedrobaeza** [comment]
Travis is red

**Pull request closed without merging**